### PR TITLE
Blender: Export Hierarchy & Transformation matrices

### PR DESCRIPTION
**Pull request by rkusa**
As of #6035 @repsac already implemented the hierarchy export, but reverted it because of transformation issues. This PR is an attempt to restore this hierarchy export, but still export transformations properly. This is done by exporting the local transformation matrix directly.

I am not used to Python, so feel free to complain about my implementation ;-)

### CODE REVIEW

File: utils/exporters/blender/addons/io_three/exporter/api/object.py [hunk 1]
```diff
@@ -395,7 +405,12 @@ def extract_mesh(obj, options, recalculate=False):
     mesh_node.calc_normals()
     mesh_node.calc_tessface()
     scale_ = options.get(constants.SCALE, 1)
-    mesh_node.transform(mathutils.Matrix.Scale(scale_, 4))
+    global_matrix = (mathutils.Matrix.Scale(scale_, 4) *
+                     axis_conversion(to_forward='-Z',
+                                     to_up='Y',
+                                     ).to_4x4())
+
```
Comment (rkusa) [inline utils/exporters/blender/addons/io_three/exporter/api/object.py]: It is not necessary to calculate the `global_matrix` for every mesh. Any suggestions where to put it instead?

Comment (rkusa) [inline utils/exporters/blender/addons/io_three/exporter/api/object.py]: Btw... is -Z as forward axis correct?

Comment (mrdoob) [inline utils/exporters/blender/addons/io_three/exporter/api/object.py]: three.js is +Z forward.

**Comment by rkusa** [comment]
Found some issues during testing. I'll report back as soon as they are fixed

**Comment by rkusa** [comment]
I've pushed a fix. I am now not able to find any other transformation issues.

**Comment by repsac** [comment]
Pulling this now. Got a test scene to run through it

**Comment by repsac** [comment]
Tested with the scene from https://github.com/mrdoob/three.js/issues/6035
image

Looks good, nice work!

**Comment by rkusa** [comment]
Should there be an option to decide between hierarchical and flat export?

**Comment by repsac** [comment]
Probably not a bad idea.

**Comment by rkusa** [comment]
Added the option and did some cleanup

screen shot 2015-03-07 at 10 58 20

File: utils/exporters/blender/addons/io_three/exporter/api/object.py [hunk 2]
```diff
@@ -278,14 +284,8 @@ def scale(obj, options):
...
     logger.debug('object.scale(%s)', obj)
-    vector = _decompose_matrix(obj)[2]
-    vector = (vector.x, vector.y, vector.z)
-
-    round_off, round_val = utilities.rounding(options)
-    if round_off:
-        vector = utilities.round_off(vector, round_val)
-
-    return vector
+    vector = matrix(obj, options).to_scale()
```
Comment (repsac) [inline utils/exporters/blender/addons/io_three/exporter/api/object.py]: Why is the rounding off of values being omitted?

**Comment by repsac** [comment]
Any specific reason why values are no longer being rounded off?

**Comment by repsac** [comment]
So just to be clear the JSON serializer is hacked to ensure it always trims float values to match the precision value set in the export. However it doesn't actually round the values up or down. The `round_off()` function handled this.

**Comment by repsac** [comment]
Ran pylint and there are a couple minor things to point out

io_three/exporter/api/object.py
```
C:260, 0: Line too long (86/80) (line-too-long)
C:248, 0: Invalid constant name "global_matrix" (invalid-name)
```

**Comment by rkusa** [comment]
Ah ok, just saw that the values are still trimed and thought it was redundant. I'll add it back and fix the pylint issues. Thanks for taking a look!

**Comment by repsac** [comment]
Concerning the rounding issues; slept on this and had a bit of an epiphany on how this should have been working. Go ahead and commit this PR and I am going to make a quick change to improve the rounding logic.

**Comment by rkusa** [comment]
Great! I've fixed the pylint issues.

**Comment by repsac** [comment]
Great work, thanks!

@mrdoob all good

**Comment by mrdoob** [comment]
Thanks guys!

**Pull request merged**